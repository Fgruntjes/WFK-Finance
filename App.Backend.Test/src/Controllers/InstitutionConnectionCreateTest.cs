using Microsoft.EntityFrameworkCore;
using Moq;
using VMelnalksnis.NordigenDotNet.Requisitions;

namespace App.Backend.Test.Controllers;

public class InstitutionConnectionCreateTest : IClassFixture<InstitutionConnectionFixture>
{
    private readonly InstitutionConnectionFixture _fixture;

    public InstitutionConnectionCreateTest(InstitutionConnectionFixture fixture)
    {
        _fixture = fixture;
    }

    [Fact]
    public async Task Success()
    {
        // Arrange
        var requisitionsMock = new Mock<IRequisitionClient>();
        _fixture.NordigenClientMoq
            .SetupGet(c => c.Requisitions)
            .Returns(requisitionsMock.Object);

        var nordigenRequisitionResult = new Requisition
        {
            Link = new Uri("https://www.example.com/connect-url/nordigen-requisition"),
            Id = new Guid("2ee63e58-5e7e-4bf2-820f-646f85876bac"),
        };
        requisitionsMock
            .Setup(r => r.Post(It.IsAny<RequisitionCreation>()))
            .ReturnsAsync(nordigenRequisitionResult);

        // Act
        var result = await _fixture.ExecuteQuery(new
        {
            InstitutionId = _fixture.InstitutionEntity.Id,
            ReturnUrl = "http://www.example.com/return"
        });
        result.MatchSnapshot();

        // Assert
        _fixture.WithData(async conext =>
        {
            var connectionEntity = await conext.InstitutionConnections
                        .Where(e => e.ExternalId == nordigenRequisitionResult.Id.ToString())
                        .FirstAsync();

            connectionEntity.Should().NotBeNull();
        });
    }

    [Fact]
    public async Task MissingInstitution()
    {
        var result = await _fixture.ExecuteQuery(new
        {
            InstitutionId = new Guid("63ee89f0-d929-4f13-9636-4efd23aee070"),
            ReturnUrl = "http://www.example.com/return"
        });
        result.MatchSnapshot(m => m.IgnoreField("errors[0].extensions.timestamp"));
    }
}