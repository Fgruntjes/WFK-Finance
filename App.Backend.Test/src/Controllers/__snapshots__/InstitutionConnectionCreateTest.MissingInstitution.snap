{
  "errors": [
    {
      "message": "Could not find institution (Parameter 'institutionId')\nActual value was 63ee89f0-d929-4f13-9636-4efd23aee070.",
      "locations": [
        {
          "line": 3,
          "column": 3
        }
      ],
      "path": [
        "institutionConnection",
        "create"
      ],
      "extensions": {
        "code": "BAD_REQUEST",
        "timestamp": "2023-09-18T19:45:48.152+02:00",
        "severity": "CRITICAL"
      }
    }
  ],
  "data": {
    "institutionConnection": {
      "create": null
    }
  }
}
